### PR TITLE
syncer: execute without retry for some error

### PR DESCRIPTION
**Pull request by nolouch**
I let all  mysql error are un-retrievable except mysql.ErrUnkonw  that alway include like "io timeout, information schema is out of date".
PTAL @qiuyesuifeng @shenli @siddontang

### CODE REVIEW

File: syncer/db.go [hunk 1]
```diff
@@ -477,6 +505,9 @@ func querySQL(db *sql.DB, query string) (*sql.Rows, error) {
 
 		rows, err = db.Query(query)
 		if err != nil {
+			if !isRetryableError(err) {
+				return rows, err
```
Comment (qiuyesuifeng) [inline syncer/db.go]: return rows, errors.Trace(err)

File: syncer/db.go [hunk 2]
```diff
@@ -516,6 +547,7 @@ LOOP:
 
 		txn, err = db.Begin()
 		if err != nil {
+			log.Warn(isRetryableError(err))
```
Comment (qiuyesuifeng) [inline syncer/db.go]: We can remove this log.

File: syncer/db.go [hunk 3]
```diff
@@ -461,6 +463,32 @@ func parserDDLTableName(sql string) (TableName, error) {
 	return res, nil
 }
 
+func isRetryableError(err error) bool {
```
Comment (qiuyesuifeng) [inline syncer/db.go]: We should add more tests for this function.

**Comment by qiuyesuifeng** [comment]
@nolouch PTAL

**Comment by nolouch** [comment]
@qiuyesuifeng  PTAL

File: syncer/db_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	return &mysql.MySQLError{
+		Number:  number,
+		Message: message,
+	}
+}
+
+func (s *testDBSuite) TestIsRetryableError(c *C) {
+	e := newMysqlErr(tmysql.ErrNoDB, "no db error")
+	r := isRetryableError(e)
+	c.Assert(r, Equals, false)
```
Comment (qiuyesuifeng) [inline syncer/db_test.go]: c.Assert(r, IsFalse)

**Comment by nolouch** [comment]
Close #44  PTAL@cwen0  @overvenus

File: syncer/db.go [hunk 5]
```diff
@@ -461,6 +463,30 @@ func parserDDLTableName(sql string) (TableName, error) {
...
+	if err == driver.ErrBadConn {
+		return true
+	}
+	var e error
+	for {
+		e = errors.Cause(err)
+		if err == e {
+			break
+		} else {
+			err = e
```
Comment (cwen0) [inline syncer/db.go]:
```
if err == e {
       break
}
err = e 
```    
is better , I think

Comment (nolouch) [inline syncer/db.go]: got it

**Comment by cwen0** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**